include/drivers/dma.h

```diff
@@ -327,6 +366,80 @@ static inline int z_impl_dma_stop(const struct device *dev, uint32_t channel)
 	return api->stop(dev, channel);
 }
 
+/**
+ * @brief request DMA channel.
+ *
+ * request DMA channel resources
+ * return -EINVAL if there is no valid channel available.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param filter_param filter function parameter
+ *
+ * @retval dma channel if successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int dma_request_channel(const struct device *dev,
+				  void *filter_param);
+
+static inline int z_impl_dma_request_channel(const struct device *dev,
+					     void *filter_param)
+{
+	int i = 0;
+	int channel = -EINVAL;
+	const struct dma_driver_api *api =
+		(const struct dma_driver_api *)dev->api;
+	/*dma_context shall be the first one in dev data*/
+	struct dma_context *dma_ctx = (struct dma_context *)dev->data;
+
+	if (dma_ctx->magic != DMA_MAGIC) {
+		return channel;
+	}
+
+	k_mutex_lock(dma_ctx->mutex, K_FOREVER);
+	for (i = 0; i < dma_ctx->dma_channels; i++) {
+		if (!atomic_test_and_set_bit(dma_ctx->atmic, i)) {
+			channel = i;
+			if (api->chan_filter &&
+			    !api->chan_filter(dev, channel, filter_param)) {
+				continue;
```

Comment:
This leaves the `atmic` bit set in a case where the channel doesn't satisfy a filter.  Is that correct?

Reply:
right, need clear if set.